src/algorithms/LaurentPoly.jl

```diff
@@ -0,0 +1,212 @@
+###############################################################################
+#
+#   LaurentPoly.jl : Generic algorithms for abstract Laurent polynomials
+#
+###############################################################################
+
+# required methods without default implementation:
+# coeff, setcoeff!, map_coeffs, gen
+
+###############################################################################
+#
+#   Data type and parent object methods
+#
+###############################################################################
+
+base_ring(p::LaurentPolyElem) = base_ring(parent(p))
+
+isdomain_type(::Type{<:LaurentPolyElem{T}}) where {T} = isdomain_type(T)
+
+isexact_type(::Type{<:LaurentPolyElem{T}}) where {T} = isexact_type(T)
+
+function check_parent(a::LaurentPolyElem, b::LaurentPolyElem, throw::Bool = true)
+   c = parent(a) == parent(b)
+   c || !throw || error("incompatible Laurent polynomial rings")
+   return c
+end
+
+
+###############################################################################
+#
+#   Basic manipulation
+#
+###############################################################################
+
+const hashp_seed = UInt === UInt64 ? 0xcdaf0e0b5ade239b : 0x5ade239b
+
+function Base.hash(p::LaurentPolyElem, h::UInt)
+   for i in monomials_degrees(p)
+      c = coeff(p, i)
+      if !iszero(c)
+         h = hash(i, h)
+         h = hash(c, h)
+      end
+   end
+   hash(hashp_seed, h)
+end
+
+# required implementation
+"""
+    monomials_degrees(p::LaurentPolyElem) -> AbstractVector{<:Integer}
+
+> Return a vector containing at least all the degrees of the non-null
+> monomials of `p`.
+"""
+function monomials_degrees end
+
+# return a monomials_degrees vector valid for both polys
+function monomials_degrees(p::LaurentPolyElem, q::LaurentPolyElem)
+   mdp = monomials_degrees(p)
+   mdq = monomials_degrees(q)
+   T = promote_type(eltype(mdp), eltype(mdq))
+   @assert T <: Signed # -one(T) must not wrap around
+   if isempty(mdq)
+      mdp, mdq = mdq, mdp
+   end
+   # if only one is empty, it's now mdp
+   if isempty(mdp) # extrema requires non-empty
+      if isempty(mdq)
+         zero(T):-one(T)
+      else
+         UnitRange{T}(extrema(mdq)...) # mdq not returned for type-stability
+      end
+   else # none is empty
+      minp, maxp = extrema(mdp)
+      minq, maxq = extrema(mdq)
+      min(minp, minq):max(maxp, maxq)
+   end
+end
+
+# like monomials_degrees, but return a "strict" range, whose min/max
+# corresponds to non-zero coeffs, or an empty range for p == 0
+function degrees_range(p::LaurentPolyElem)
+   mds = monomials_degrees(p)
+   T = eltype(mds)
+   minp, maxp = isempty(mds) ? (T(0), T(0)) : extrema(mds)
+   while iszero(coeff(p, minp))
+      minp += 1
+      minp > maxp && return minp:maxp
+   end
+   while iszero(coeff(p, maxp))
+      maxp -= 1
+   end
+   minp:maxp
+end
+
+
+# return the degree of the unique monomial, throw if not a monomial
+function monomial_degree(p::LaurentPolyElem)
```

Comment:
What would be an example of an application for this function?

Reply:
There is only one use of this function in the implementation of `^` ([see here](https://github.com/Nemocas/AbstractAlgebra.jl/pull/593/files#diff-a7d0334567eeacdbc32f88494465e6e2R127)), so for now this can be considered as an internal function